File: scala/P12.scala
// P12 Decode a run-length encoded list.

// List of objects
def f1(list: List[Any]): List[Any] = list.flatMap {
  case (n: Int, x) => List.fill(n)(x)
  case x => x :: Nil
}

// With custom data
abstract sealed class Entry[T]
case class Value[T](v: T) extends Entry[T]
case class Sequence[T](n: Int, v: T) extends Entry[T]

def f2[T](list: List[Entry[T]]): List[T] = list.flatMap {
  case Value(v) => v :: Nil
  case Sequence(n, v) => List.fill(n)(v)
}

File: haskell/p12.hs
-- P12 Decode a run-length encoded list.

data Entry a = Value a | Sequence Int a deriving Show

-- Using concatMap
f1 :: [Entry a] -> [a]
f1 = concatMap toItems
    where toItems (Value v) = [v]
          toItems (Sequence n v) = replicate n v

-- Recursion
f2 :: [Entry a] -> [a]
f2 [] = []
f2 ((Value v) : xs) = v : f2 xs
f2 ((Sequence n v) : xs) = replicate n v ++ f2 xs